packages/alfa-rules/src/sia-er62/rule.ts

```diff
@@ -69,54 +61,111 @@ export default Rule.Atomic.of<Page, Element>({
-                and(hasRole(device, "paragraph"), hasNonLinkText(device)),
-                node
-              )
-            ) {
+            if (isParagraph(node)) {
               container = Option.of(node);
             }
           }
 
+          const isTextNode = test(and(isText, isVisible(device)), node);
+          const parent = node.parent().filter(isElement);
+          if (isTextNode && container.isSome() && parent.isSome()) {
+            // For each link, store the parent of the text nodes it contains
+            if (link.isSome()) {
+              linkText = linkText.set(
+                link.get(),
+                linkText
+                  .get(link.get())
+                  .getOr(Set.empty<Element>())
+                  .add(parent.get())
+              );
+            }
+            // For each container, store the parent of the text nodes it contains
+            else {
+              nonLinkText = nonLinkText.set(
+                container.get(),
+                nonLinkText
+                  .get(container.get())
+                  .getOr(Set.empty<Element>())
+                  .add(parent.get())
+              );
+            }
+          }
+
           const children = node.children({
             flattened: true,
             nested: true,
           });
 
           for (const child of children) {
-            yield* visit(child, container);
+            gather(child, container, link);
+          }
+        }
+
+        function* getApplicableLinks(): Iterable<Element> {
+          // Check if foreground is the same with the parent <p> element
+          const hasDifferentForeground = (
+            link: Element,
+            container: Element
+          ): boolean =>
+            getForeground(link, device).none(
+              (linkColors) =>
+                // If the link has a foreground with the alpha channel less than 1 and background gradient color
+                // then the rule is applicable as we can't tell for sure if it ever has the same foreground with a link
+                // that might have the same foreground and gradien background, but with different gradient type or spread
+                linkColors.length === 1 &&
+                getForeground(container, device).some(
+                  (containerColors) =>
+                    // Similalry to the link, we assume it's applicable if the container has different foreground colors
+                    containerColors.length === 1 &&
+                    linkColors[0].equals(containerColors[0])
+                )
+            );
+
+          for (const [link, linkTexts] of linkText) {
+            const nonLinkTexts = nonLinkText
+              .get(containers.get(link).get())
+              .get();
+
+            for (const linkElement of linkTexts) {
+              for (const nonLinkElement of nonLinkTexts) {
+                if (hasDifferentForeground(linkElement, nonLinkElement)) {
+                  return yield link;
```

Comment:
```suggestion
            loop:
            for (const linkElement of linkTexts) {
              for (const nonLinkElement of nonLinkTexts) {
                if (hasDifferentForeground(linkElement, nonLinkElement)) {
                  yield link;
                  break loop;
```
should work.

or something in the line of
```suggestion
            if (linkTexts.some(linkElement =>
              nonLinkTexts.some(nonLinkElement =>
                hasDifferentForeground(linkElement, nonLinkElement))))
                  {yield link;}
```
The the lazyness of `.some` should take care of everything for us.